mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,644 @@
+    InteractionState,
+    LearnerState,
+    Metrics,
+    TrainState,
+    Transition,
+)
+from mava.types import Observation
+from mava.utils import make_env as environments
+from mava.utils.checkpointing import Checkpointer
+from mava.utils.jax import unreplicate_batch_dim, unreplicate_n_dims
+from mava.utils.logger import LogEvent, MavaLogger
+from mava.wrappers import episode_metrics
+
+
+def init(
+    cfg: DictConfig,
+) -> Tuple[
+    Tuple[Environment, Environment],
+    RecQNetwork,
+    optax.GradientTransformation,
+    TrajectoryBuffer,
+    LearnerState,
+    MavaLogger,
+    PRNGKey,
+]:
+    """Initialize system by creating the envs, networks etc.
+
+    Args:
+        cfg: System configuration.
+
+    Returns:
+        Tuple containing:
+            Tuple[Environment, Environment]: The environment and evaluation environment.
+            RecQNetwork: Recurrent Q network
+            optax.GradientTransformation: Optimiser for RecQNetwork.
+            TrajectoryBuffer: The replay buffer.
+            LearnerState: The initial learner state.
+            MavaLogger: The logger.
+            PRNGKey: The random key.
+    """
+    logger = MavaLogger(cfg)
+
+    key = jax.random.PRNGKey(cfg.system.seed)
+    devices = jax.devices()
+
+    def replicate(x: Any) -> Any:
+        """First replicate the update batch dim then put on devices."""
+        x = jax.tree_map(lambda y: jnp.broadcast_to(y, (cfg.system.update_batch_size, *y.shape)), x)
+        return jax.device_put_replicated(x, devices)
+
+    env, eval_env = environments.make(cfg)
+
+    action_dim = env.action_dim
+    num_agents = env.num_agents
+
+    key, q_key = jax.random.split(key, 2)
+
+    # Make dummy inputs to init recurrent Q network -> need format TBAx
+    init_obs = env.observation_spec().generate_value()  # A,x
+    init_obs_batched = jax.tree_util.tree_map(
+        lambda x: jnp.repeat(x[jnp.newaxis, ...], cfg.arch.num_envs, axis=0),
+        init_obs,
+    )  # B, A , x
+    init_obs_batched = jax.tree_util.tree_map(
+        lambda x: x[jnp.newaxis, ...], init_obs_batched
+    )  # add time dim (1,B,A,x)
+    init_done = jnp.zeros((1, cfg.arch.num_envs, 1), dtype=bool)  # (1,B,1)
+    init_x = (init_obs_batched, init_done)  # pack the RNN dummy inputs
+    init_hidden_state = ScannedRNN.initialize_carry(
+        (cfg.arch.num_envs, num_agents), cfg.network.hidden_state_dim
+    )  # (B, A, x)
+
+    # Making recurrent Q network
+    q_net = RecQNetwork(action_dim, cfg.network.hidden_state_dim)
+    q_params = q_net.init(q_key, init_hidden_state, init_x)  # epsilon defaults to 0
+    q_target_params = q_net.init(q_key, init_hidden_state, init_x)  # ensure parameters are separate
+
+    # Pack Q network params
+    params = DDQNParams(q_params, q_target_params)
+
+    # Making optimiser and state
+    opt = optax.chain(
+        # optax.clip_by_global_norm(10), # used in JAXMARL
+        optax.adam(learning_rate=cfg.system.q_lr),  # eps=1e-5 in JAXMARL
```

Comment:
Is it useful to keep the clipping "in reserve" like this for comparison to JAXMARL, or is it okay without?

Reply:
I think add clipping in and add it to the config

Reply:
I think we can set `eps` to be `1e-5` by default and leave it as is. We can also keep the `optax.clip_by_global_norm` and set this using a parameter we specify in the system configs similarly to what we do in the ppo systems.